sdks/go/pkg/beam/io/mongodbio/write.go

```diff
@@ -0,0 +1,204 @@
+// that field will be used as the id of the document. Otherwise, a new id field of type
+// primitive.ObjectID will be generated for each document. Write returns a PCollection<K> of the
+// inserted id values with type K.
+//
+// The Write transform has the required parameters:
+// 	- s: the scope of the pipeline
+// 	- uri: the MongoDB connection string
+// 	- database: the MongoDB database to write to
+// 	- collection: the MongoDB collection to write to
+// 	- col: the PCollection to write to MongoDB
+//
+// The Write transform takes a variadic number of WriteOptionFn which can set the WriteOption
+// fields:
+// 	- BatchSize: the number of documents to write in a single batch. Defaults to 1000
+// 	- Ordered: whether to execute the writes in order. Defaults to true
+func Write(
+	s beam.Scope,
+	uri string,
+	database string,
+	collection string,
+	col beam.PCollection,
+	opts ...WriteOptionFn,
+) beam.PCollection {
+	s = s.Scope("mongodbio.Write")
+
+	option := &WriteOption{
+		BatchSize: defaultWriteBatchSize,
+		Ordered:   defaultWriteOrdered,
+	}
+
+	for _, opt := range opts {
+		if err := opt(option); err != nil {
+			panic(fmt.Sprintf("mongodbio.Write: invalid option: %v", err))
+		}
+	}
+
+	t := col.Type().Type()
+	idIndex := structx.FieldIndexByTag(t, bsonTag, "_id")
+
+	var keyed beam.PCollection
+
+	if idIndex == -1 {
+		pre := beam.ParDo(s, createIDFn, col)
+		keyed = beam.Reshuffle(s, pre)
+	} else {
+		keyed = beam.ParDo(
+			s,
+			newExtractIDFn(idIndex),
+			col,
+			beam.TypeDefinition{Var: beam.XType, T: t.Field(idIndex).Type},
+		)
+	}
+
+	return beam.ParDo(
+		s,
+		newWriteFn(uri, database, collection, option),
+		keyed,
+	)
+}
+
+func createIDFn(elem beam.Y) (primitive.ObjectID, beam.Y) {
+	id := primitive.NewObjectID()
+	return id, elem
+}
+
+type extractIDFn struct {
+	IDIndex int
+}
+
+func newExtractIDFn(idIndex int) *extractIDFn {
+	return &extractIDFn{
+		IDIndex: idIndex,
+	}
+}
+
+func (fn *extractIDFn) ProcessElement(
+	_ context.Context,
+	elem beam.Y,
+	emit func(beam.X, beam.Y),
+) {
+	id := reflect.ValueOf(elem).Field(fn.IDIndex).Interface()
+	emit(id, elem)
+}
+
+type writeFn struct {
+	mongoDBFn
+	BatchSize int64
+	Ordered   bool
+	models    []mongo.WriteModel
+}
+
+func newWriteFn(
+	uri string,
+	database string,
+	collection string,
+	option *WriteOption,
+) *writeFn {
+	return &writeFn{
+		mongoDBFn: mongoDBFn{
+			URI:        uri,
+			Database:   database,
+			Collection: collection,
+		},
+		BatchSize: option.BatchSize,
+		Ordered:   option.Ordered,
+	}
+}
+
+func (fn *writeFn) ProcessElement(
+	ctx context.Context,
+	key beam.X,
+	value beam.Y,
+	emit func(beam.X),
+) error {
+	model := mongo.NewReplaceOneModel().
+		SetFilter(bson.M{"_id": key}).
+		SetUpsert(true).
+		SetReplacement(value)
+
+	fn.models = append(fn.models, model)
+
+	if len(fn.models) >= int(fn.BatchSize) {
+		if err := fn.flush(ctx); err != nil {
+			return err
+		}
+	}
+
+	emit(key)
```

Comment:
Having the same concern about exposing third party libraries here where ids are emitted. If the element struct did not have an id field, the id would have been generated with `createIDFn`, and with the current implementation that would be a `primitive.ObjectID`. Is it preferred to emit a `[12]byte` instead? (If the user themselves defined a struct with an id of type primitive.ObjectID, i.e. the id is extracted with `extractIDFn`, I think the use case is different and they should expect to get a primitive.ObjectID back)

Reply:
Regarding exposing 3rd party libraries, the concern is typically on the user surface. So internal to a composite AKA, all the DoFns + PCollections that are being put together on the pipeline when they call `Read` or `Write`, that's fine. The only concern is ensuring the type is encodeable (as all elements in PCollections must be).

Un-encodeable types can slip in for direct runners, or runners with aggressive fusion, but shouldn't be taken as given for execution.

In particular, we don't want DoFns to agggresively serialize things themselves to pass between other DoFns, since that's inefficient, and avoiding that inefficiency is why fusion exists as an optimization, and why Beam is very Coder Aware.